java-checks/src/main/java/org/sonar/java/checks/EqualsArgumentTypeCheck.java

```diff
@@ -112,46 +152,36 @@ public void visitBinaryExpression(BinaryExpressionTree tree) {
     @Override
     public void visitMethodInvocation(MethodInvocationTree tree) {
       if (EQUALS_MATCHER.matches(tree)) {
-        typeChecked = true;
+        if (isInvocationOnArgument(tree)) {
+          typeChecked = true;
+        } else if (isArgument(tree.arguments().get(0), parameterSymbol)) {
+          typeChecked = true;
+        }
       }
     }
 
-    private boolean isArgument(ExpressionTree tree) {
-      ExpressionTree expressionTree = removeParenthesis(tree);
-      return expressionTree.is(Tree.Kind.IDENTIFIER) && ((IdentifierTree) expressionTree).symbol().equals(parameterSymbol);
-    }
-
     private boolean isExplicitComparison(ExpressionTree operand1, ExpressionTree operand2) {
-      return (isArgument(operand1) && isThis(operand2)) || (isArgument(operand2) && isThis(operand1));
+      return (isArgument(operand1, parameterSymbol) && isThis(operand2)) || (isArgument(operand2, parameterSymbol) && isThis(operand1));
     }
 
     private boolean isGetClassOnArgument(ExpressionTree tree) {
       ExpressionTree expressionTree = removeParenthesis(tree);
       if (expressionTree.is(Tree.Kind.METHOD_INVOCATION) && GETCLASS_MATCHER.matches((MethodInvocationTree) expressionTree)) {
-        ExpressionTree methodSelect = ((MethodInvocationTree) expressionTree).methodSelect();
-        if (methodSelect.is(Tree.Kind.MEMBER_SELECT)) {
-          ExpressionTree expression = ((MemberSelectExpressionTree) methodSelect).expression();
-          return isArgument(expression);
+        if (isInvocationOnArgument((MethodInvocationTree) expressionTree)) {
```

Comment:
:new: ![MAJOR](https://raw.githubusercontent.com/henryju/image-hosting/master/major.png)Merge this if statement with the enclosing one.[![...](https://raw.githubusercontent.com/henryju/image-hosting/master/more.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1066)